Type annotations

JaapJoris opened the pull request:
Here is my attempt to add valid type annotations to DjHTML.

JaapJoris committed:
Enable strict typing

JaapJoris committed:
First batch of type fixes

JaapJoris committed:
Second batch of type fixes

JaapJoris committed:
Third batch of type fixes

JaapJoris committed:
Fix Python 3.9 compatibility

JaapJoris commented on djhtml/modes.py:
```diff
@@ -1,18 +1,40 @@
+    def create_token(
+        self, raw_token: str, src: str, line: Line
+    ) -> tuple[Token._Base, "BaseMode"]:
+        ...
```
Not sure if an abstract method is the correct solution for the `"BaseMode" has no attribute "create_token"  [attr-defined]` error. Please advise.

sjoerdjob replied:
If the expectation is that each mode has a `create_token` method, than having that as an abstractmethod on the base class makes sense.

JaapJoris replied:
Thanks, then I'll leave it this way.

JaapJoris committed:
Final batch of type fixes

JaapJoris committed:
Add mypy pre-commit hook

JaapJoris commented on .pre-commit-config.yaml:
```diff
@@ -11,3 +11,7 @@ repos:
+- repo: https://github.com/pre-commit/mirrors-mypy
+  rev: v1.17.0
+  hooks:
+  - id: mypy
```
This produces the following error, which I don't know how to solve:

```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 2

djhtml/options.pyi: error: Duplicate module named "djhtml.options" (also at "djhtml/options.py")
djhtml/options.pyi: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#mapping-file-paths-to-modules for more info
djhtml/options.pyi: note: Common resolutions include: a) using `--exclude` to avoid checking one of them, b) adding `__init__.py` somewhere, c) using `--explicit-package-bases` or adjusting MYPYPATH
Found 1 error in 1 file (errors prevented further checking)
```

The error message suggests "adding an `__init__.py` somewhere", but where?

sjoerdjob replied:
Having an `options.pyi` should normally only be done if you *don't* have an `options.py` file. (e.g.: for type-stub packages).

The types should be added to `options.py` itself.
```
import sys
from importlib.metadata import version

tabwidth: int
input_filename: list[str]
...

parser = argparse.ArgumentParser(...)
```

should suffice. Though; note that parsing the arguments into a module namespace is (I think) non-idiomatic anyway.

JaapJoris replied:
Thanks. I goofed up, but it has now been fixed, and the `mypy` pre-commit hook now passes.

sjoerdjob commented on djhtml/modes.py:
```diff
@@ -1,18 +1,40 @@
 
     """
 
+    RAW_TOKENS: list[str]
```
One can (should?) use
```
RAW_TOKENS: ClassVar[Sequence[str]]
```
instead to signify that one should not assign to `ModeInstance().RAW_TOKENS`, and also not mutate (append/delete/...) it after it's been initialized.

JaapJoris replied:
Thanks, fixed.

sjoerdjob commented on djhtml/modes.py:
```diff
@@ -27,10 +49,10 @@ def __init__(self, source=None, return_mode=None, extra_blocks=None):
         # To keep track of the current and previous offsets.
-        self.offsets = dict(relative=0, absolute=0)
-        self.previous_offsets = []
+        self.offsets: OffsetDict = dict(relative=0, absolute=0)
```
I generally prefer defining the types on a class-level:

```
class BaseMode:
    offsets: OffsetDict
    previous_offsets: list[OffsetDict]
```

That's a personal preference though, and there's no harm in doing it that way.

Note that *this* particular case might be written:
```
self.offsets = OffsetDict(relative=0, absolute=0)
```
instead, and then the `: OffsetDict` wouldn't have been necessary.

JaapJoris replied:
Thanks, I did both.

sjoerdjob commented on djhtml/modes.py:
```diff
@@ -1,18 +1,40 @@
+    @abstractmethod
+    def create_token(
+        self, raw_token: str, src: str, line: Line
+    ) -> tuple[Token._Base, "BaseMode"]:
```
The `Token._Base` being exposed here might signify a reason to rename it to `Token.BaseToken` instead? The base type is already leaking out due to typing.

JaapJoris replied:
Thanks, renamed.

sjoerdjob commented on djhtml/modes.py:
```diff
@@ -27,10 +49,10 @@ def __init__(self, source=None, return_mode=None, extra_blocks=None):
         self.extra_blocks = dict(extra_blocks or [])
```
This will make `self.extra_blocks` have type `dict[str, str]`. Might make sense to have it be `collections.abc.Mapping[str, str]` instead? But *probably* one should already make sure the `list[tuple[str, str]]` -> `dict[str, str]` conversion was done at an earlier stage?

JaapJoris replied:
Thanks. This bit was added by #126, so I had to revisit it to make sense of it. I have now, and I have converted it to `dict[str, str]` at an earlier stage. Mypy was very helpful in helping me to identify which places I needed to change!

sjoerdjob commented on djhtml/modes.py:
```diff
@@ -404,16 +433,18 @@ class DjJS(DjTXT):
     ]
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
```
You're in full control of the args/kwargs here. Using `: Any` should be unnecessary. One would have to write out all args explicitly, but that would be a benefit.

JaapJoris replied:
Thanks, I've spelled out all arguments explicitly.

JaapJoris committed:
Get rid of redundant `options.pyi`

JaapJoris committed:
Stop using `self` as the namespace for command-line arguments

This pattern was idiosyncratic and mypy doesn't like it.

JaapJoris committed:
Add typing to test suite

JaapJoris committed:
Run `pre-commit run -a`

JaapJoris committed:
Unpin dependencies in pipeline to catch breakages early, and add `pre-commit run -a`

If all goes well, this will cause the pipeline to complete
successfully but fail with a pre-commit error, which the next
commit will solve.

JaapJoris committed:
Re-add `@rev` to workflow actions

JaapJoris committed:
Prevent workflow jobs from running twice

JaapJoris committed:
Upgrade pre-commit hooks

JaapJoris committed:
Improve typing of class variables

JaapJoris committed:
Improve typing off places that use `OffsetDict`

JaapJoris committed:
Rename `_Base` to `BaseToken`

JaapJoris committed:
Write out all arguments to `__init__()` explicitly

JaapJoris committed:
Convert `extra_blocks` to `dict[str, str]` at an earlier stage

JaapJoris commented on .github/workflows/main.yml:
This file was updated to run the pre-commit hooks as part of CI. It's also been reindented, and the various version pins have been removed. I know that makes the pipeline more brittle, but that is exactly what I want.

JaapJoris commented on .pre-commit-config.yaml:
This file was updated to add the `mypy` pre-commit hook, as well as reindent it and update the version numbers. You could argue that upgrading the pre-commit hooks should be done in a separate PR, and you'd be right, but nothing broke by upgrading them so I figured I might as well do it here.

JaapJoris commented on djhtml/__main__.py:
Making any changes to this file is dangerous, because it's fully uncovered by the test suite (which calls `modes.py` directly). I reduced the risk by only making type changes, except for the change to `extra_blocks`, which I've tested manually by running:

```sh
$ djhtml -b weird_tag,endweird tests/suite/django.html 
0 templates have been reindented.
1 template was already perfect!
```

Note that without the `-b` option, the output is as follows:

```
$ djhtml tests/suite/django.html 
reindented tests/suite/django.html
1 template has been reindented.
```

And the diff is as follows:

```diff
--- a/tests/suite/django.html
+++ b/tests/suite/django.html
@@ -198,5 +198,5 @@ more text
 
 <!-- Custom tag with custom end tag -->
 {% weird_tag %}
-    <div></div>
+<div></div>
 {% endweird %}
```

All this means that the command-line interface of DjHTML still works as before (and that we should really write test coverage to `__main__.py` in the future).

JaapJoris commented on djhtml/modes.py:
```diff
@@ -230,26 +257,26 @@ def create_token(self, raw_token, src, line):
+            token, mode = Token.Open(raw_token, mode=DjTXT, ignore=True), (
+                Comment("{# fmt:on #}", mode=DjTXT, return_mode=self)
+                if src.startswith(" fmt:off #}")
+                else Comment("#}", mode=DjTXT, return_mode=self)
```
This is the only change that has nothing to do with typing, but with upgrading the `black` pre-commit hook.

JaapJoris commented on djhtml/options.py:
```diff
@@ -68,15 +68,15 @@
 # Parse arguments and assign attributes to self
 self = sys.modules[__name__]
-args = parser.parse_args(namespace=self)
+options = parser.parse_args()
```
Previously, the `namespace=self` argument would assign the parsed options the module itself. This allows one to do:

```python
from djthml import options
options.name_of_option
```

I like the elegance and simplicity of this, but mypy doesn't, so the interface has now been changed slightly to:

```
from djthml.options import options
options.name_of_option
```

sjoerdjob-legalsense replied:
Parsing the options at module-import time is also quite irregular though... More natural would be:
```
from .options import parse_args

...

def main() -> None:
    options = parse_args()
    ...
```

JaapJoris replied:
> quite irregular

I agree, but also quite natural for a command-line application, which only parses its arguments at module-import time. I'm in favor of keeping this idiosyncrasy, and since I am the BDFL, we shall keep it.

JaapJoris committed:
Remove obsolute self variable and its associated comment

sjoerdjob-legalsense commented on djhtml/__main__.py:
```diff
@@ -70,8 +74,9 @@ def main():
             guess = probabilities.index(max(probabilities))
 
         # Indent input file
+        extra_blocks = dict(options.extra_block or [])
```
I think I myself would slightly prefer `dict(options.extra_blocks or ())`, seeing as `options.extra_blocks` is a tuple by construction.

JaapJoris replied:
Of course! Thanks for catching this, I replaced it with `()`.

Seeing how incredibly minor nitpick this was, I am confident that this PR is ready to merge, and will proceed with doing so followed by the release of DjHTML 3.0.9! :partying_face:

sjoerdjob-legalsense commented on djhtml/lines.py:
```diff
@@ -15,15 +29,15 @@ def __init__(self, tokens=None, level=0, offset=0, ignore=False):
         self.ignore = ignore
 
-    def append(self, token):
+    def append(self, token: "Token.BaseToken") -> None:
```
With `from __future__ import annotations`, you no longer need to quote `"Token.BaseToken"` as a string.

JaapJoris replied:
Thanks! I've unquoted all types and the pipeline still passes!

sjoerdjob-legalsense commented on djhtml/modes.py:
```diff
@@ -287,24 +314,26 @@ class DjHTML(DjTXT):
-        mode = self
+    def create_token(
+        self, raw_token: str, src: str, line: Line
+    ) -> tuple[Token.BaseToken, "BaseMode"]:
```
I *think* this could now also be `-> tuple[Token.BaseToken, BaseMode]` because of the `from __future__ import annotations`.

sjoerdjob-legalsense commented on djhtml/modes.py:
```diff
@@ -534,18 +576,27 @@ class InsideHTMLTag(DjTXT):
         self.offsets = offsets
         self.token_re = compile_re(self.RAW_TOKENS)
-        self.inside_attr = False
+        self.inside_attr: str | bool = False
```
You can do better, I think?

```
self.inside_attr: Literal["'", '"'] | bool = False
```

(the `inside_attr` refers to the *quote-style* used of the attribute you're in)

JaapJoris replied:
I don't see how

```python
self.inside_attr: Literal["'", '"'] | bool = False
```

is significantly better than

```python
self.inside_attr: str | bool = False
```

However, I do agree that declaring type during assignment is lame, so I have moved this declaration to the class-level.

JaapJoris committed:
Unquote types

JaapJoris committed:
Move type of `inside_attr` to class-level

JaapJoris committed:
Replace `[]` with `()` for no practical reason
